lib/manifest.go

```diff
@@ -0,0 +1,198 @@
+package lib
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// Manifest stores Kubernetes object in Raw from a file.
+// It stores the GroupVersionKind for the manifest.
+type Manifest struct {
+	Raw []byte
+	GVK schema.GroupVersionKind
+
+	obj *unstructured.Unstructured
+}
+
+// UnmarshalJSON unmarshals bytes of single kubernetes object to Manifest.
+func (m *Manifest) UnmarshalJSON(in []byte) error {
+	if m == nil {
+		return errors.New("Manifest: UnmarshalJSON on nil pointer")
+	}
+
+	// This happens when marshalling
+	// <yaml>
+	// ---	(this between two `---`)
+	// ---
+	// <yaml>
+	if bytes.Equal(in, []byte("null")) {
+		m.Raw = nil
+		return nil
+	}
+
+	m.Raw = append(m.Raw[0:0], in...)
+	udi, _, err := scheme.Codecs.UniversalDecoder().Decode(in, nil, &unstructured.Unstructured{})
+	if err != nil {
+		return fmt.Errorf("unable to decode manifest: %v", err)
+	}
+	ud, ok := udi.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("expected manifest to decode into *unstructured.Unstructured, got %T", ud)
+	}
+
+	m.GVK = ud.GroupVersionKind()
+	m.obj = ud.DeepCopy()
+	return nil
+}
+
+// Object returns underlying metav1.Object
+func (m *Manifest) Object() metav1.Object { return m.obj }
+
+const (
+	// rootDirKey is used as key for the manifest files in root dir
+	// passed to LoadManifests
+	// It is set to `000` to give it more priority if the actor sorts
+	// based on keys.
+	rootDirKey = "000"
```

Comment:
Alternatively we could just force the root dir to be a real dir and ignore files in the root (treat as not manifests).  The only content we had in there in the payload right now was the image mapping and the Cincinnati file.

Reply:
If the update payload is 
```console
/cincinnati.json
/images.json
manifests/
   ....
   ....
```
The root here would mean manifests dir

Reply:
+1 on keeping the manifests dir, seems pretty clean to me and not much overhead.